tiledb/sm/array_schema/dimension.cc

```diff
@@ -2114,972 +1610,112 @@ void Dimension::set_compute_mbr_func() {
-    case Datatype::DATETIME_WEEK:
-    case Datatype::DATETIME_DAY:
-    case Datatype::DATETIME_HR:
-    case Datatype::DATETIME_MIN:
-    case Datatype::DATETIME_SEC:
-    case Datatype::DATETIME_MS:
-    case Datatype::DATETIME_US:
-    case Datatype::DATETIME_NS:
-    case Datatype::DATETIME_PS:
-    case Datatype::DATETIME_FS:
-    case Datatype::DATETIME_AS:
-    case Datatype::TIME_HR:
-    case Datatype::TIME_MIN:
-    case Datatype::TIME_SEC:
-    case Datatype::TIME_MS:
-    case Datatype::TIME_US:
-    case Datatype::TIME_NS:
-    case Datatype::TIME_PS:
-    case Datatype::TIME_FS:
-    case Datatype::TIME_AS:
-      relevant_ranges_func_ = relevant_ranges<int64_t>;
-      break;
-    case Datatype::STRING_ASCII:
+  auto g = [&](auto T) {
+    if constexpr (std::is_same_v<decltype(T), char>) {
       assert(var_size());
-      relevant_ranges_func_ = relevant_ranges<char>;
-      break;
-    default:
-      relevant_ranges_func_ = nullptr;
-      break;
-  }
+    }
+    relevant_ranges_func_ = relevant_ranges<decltype(T)>;
+  };
+  execute_callback_with_type(type_, g);
 }
 
 void Dimension::set_covered_vec_func() {
-  switch (type_) {
-    case Datatype::INT32:
-      covered_vec_func_ = covered_vec<int32_t>;
-      break;
-    case Datatype::INT64:
-      covered_vec_func_ = covered_vec<int64_t>;
-      break;
-    case Datatype::INT8:
-      covered_vec_func_ = covered_vec<int8_t>;
-      break;
-    case Datatype::UINT8:
-      covered_vec_func_ = covered_vec<uint8_t>;
-      break;
-    case Datatype::INT16:
-      covered_vec_func_ = covered_vec<int16_t>;
-      break;
-    case Datatype::UINT16:
-      covered_vec_func_ = covered_vec<uint16_t>;
-      break;
-    case Datatype::UINT32:
-      covered_vec_func_ = covered_vec<uint32_t>;
-      break;
-    case Datatype::UINT64:
-      covered_vec_func_ = covered_vec<uint64_t>;
-      break;
-    case Datatype::FLOAT32:
-      covered_vec_func_ = covered_vec<float>;
-      break;
-    case Datatype::FLOAT64:
-      covered_vec_func_ = covered_vec<double>;
-      break;
-    case Datatype::DATETIME_YEAR:
-    case Datatype::DATETIME_MONTH:
-    case Datatype::DATETIME_WEEK:
-    case Datatype::DATETIME_DAY:
-    case Datatype::DATETIME_HR:
-    case Datatype::DATETIME_MIN:
-    case Datatype::DATETIME_SEC:
-    case Datatype::DATETIME_MS:
-    case Datatype::DATETIME_US:
-    case Datatype::DATETIME_NS:
-    case Datatype::DATETIME_PS:
-    case Datatype::DATETIME_FS:
-    case Datatype::DATETIME_AS:
-    case Datatype::TIME_HR:
-    case Datatype::TIME_MIN:
-    case Datatype::TIME_SEC:
-    case Datatype::TIME_MS:
-    case Datatype::TIME_US:
-    case Datatype::TIME_NS:
-    case Datatype::TIME_PS:
-    case Datatype::TIME_FS:
-    case Datatype::TIME_AS:
-      covered_vec_func_ = covered_vec<int64_t>;
-      break;
-    case Datatype::STRING_ASCII:
+  auto g = [&](auto T) {
+    if constexpr (std::is_same_v<decltype(T), char>) {
       assert(var_size());
```

Comment:
I think this assert could be changed to throwing... It's not like set_covered_vec_func will get called in any performance dependent code.

Reply:
There are a few other places that would also need to be fixed in this file.